File: swift/binary-search-tree/Sources/BinarySearchTree/BinarySearchTree.swift
public class BinarySearchTree : Equatable {
    var left: BinarySearchTree? = nil
    var right: BinarySearchTree? = nil
    var data: Int
    
// inits:
    init(_ data: Int) { self.data = data }

// instance methods:
    func insert(_ x: Int) {
        if x > data {
            if right != nil { right!.insert(x) }
            else if right != self { right = BinarySearchTree(x) }
        } else if left != self { // x <= data
            if left != nil { left?.insert(x) }
            else { left = BinarySearchTree(x) }
        }
    }
    
    func allData() -> [Int] {
        return (left?.allData() ?? []) + [data] + (right?.allData() ?? [])
    }

// operators:
    public static func == (a: BinarySearchTree, b: BinarySearchTree) -> Bool {
        return a.data == b.data
    }
}

File: cpp/binary-search-tree/binary_search_tree.cpp
#include "binary_search_tree.h"

namespace binary_search_tree {

template <class T> auto binary_tree<T>::insert(const T value) noexcept {
    if (value > data_) {
      if (right_) {
        if (right_.get() != this) right_->insert(value);
        else                      right_ = Branch(new T(value));
      } else if (left_.get() != this) {
        if (left_)                left_->insert(value);
        else                      left_ = Branch(new T(value));
      }
    }
 }
 template <class T> auto binary_tree<T>::right() const noexcept -> Branch  {
   return right_;
 }
 template <class T> auto binary_tree<T>::left() const noexcept -> Branch {
   return left_;
 }
 template <class T> auto binary_tree<T>::data() noexcept -> T& {
   return data_;
 }
 
} // namespace binary_search_tree